### ISSUE TITLE
What exactly are the units for the force output on the test_image.png when running the force estimator?

### ISSUE THREAD

**Issue by f5ces27er5c767s** [linchangyi1/9DTact#9]
Hi @linchangyi1 thank you for your strong guidance and help,

I also wanted to start a new github thread and ask about this output we got when testing the force estimate when running estimator.py:

![to-send4](https://github.com/user-attachments/assets/d90e50ad-4350-4e11-8b27-e4d083b77b37)

Please just focus in on the force estimate output: [0.5228931  0.47465006 0.73858875 0.46580032 0.53931546 0.4671362]

I wanted to ask, what units is each element in? Is it Newtons (N) or milli-Newtons? and what about the torques?

Before the 3-step calibration process we tried testing the model accuracy on a single image whose associated force exertion we know (from our own ground-truth load cell) but the results were greatly different, and maybe it is because the units are not the same. Is the 3-step camera calibration process essential for accurate force-estimate?

Thank you @linchangyi1 kindly for all your help! We are excited to get this sensorr working!!

**Comment by linchangyi1** [comment]
The output is the 6d normalized force estimation which is in the range of [0, 1]. Please check 'wrench_range' in https://github.com/linchangyi1/9DTact/blob/main/data_collection/wrench_normalization.py for real scale. 
The calibration steps are just for shape reconstruction. Please read the paper and the code for more details.

**Comment by f5ces27er5c767s** [comment]
Hi @linchangyi1 thanks for messaging and replying, I see about the normalized force estimation output. I modified the code below so as to also output the unnormalized real scale values of the force. I am having trouble understanding why on a test image with no force applied to it at all, the reading in Fz gives a nonzero value, could you help us with this? :)

My modified estimator.py is:

```
import os
import sys

sys.path.append(os.path.dirname(os.path.abspath(__file__)) + '/../')
import yaml
import torch
import numpy as np
np.set_printoptions(suppress=True)
import argparse
from model import *
import cv2

class Estimator:
    def __init__(self, cfg):
        parser = argparse.ArgumentParser()
        parser.add_argument("--model_name", default=None, type=str)
        parser.add_argument("--model_layer", default=None, type=str)
        parser.add_argument("--cuda_index", default=None, type=int)
        args = parser.parse_args()

        # parameters
        model_name = args.model_name if args.model_name is not None else cfg['model_list'][cfg['model_choice']][0]
        model_layer = args.model_layer if args.model_name is not None else cfg['model_list'][cfg['model_choice']][1]
        model_type = model_name + '-' + str(model_layer)
        cuda_index = args.cuda_index if args.cuda_index is not None else cfg['cuda_index']

        # model configuration
        self.device = torch.device(f"cuda:{cuda_index}" if torch.cuda.is_available() else "cpu")
        if model_name == 'Resnet':
            self.model = Resnet(layer=int(model_layer), pretrained=False).to(self.device)
        elif model_name == 'Densenet':
            self.model = Densenet(layer=int(model_layer), pretrained=False).to(self.device)

        self.save_dir = cfg['save_dir'] + '/' + model_type + str(cfg['weights'][cfg['model_choice']][0])
        weights_path = self.save_dir + '/epoch_' + str(cfg['weights'][cfg['model_choice']][1]) + '.pt'
        self.model.load_state_dict(torch.load(weights_path, map_location=self.device))
        self.model.eval()
        # inferring the first image takes more time than the latter images, so we pass it
        with torch.no_grad():
            two_img = np.ones((2, 3, cfg['img_size'][0], cfg['img_size'][1]))
            representation = torch.from_numpy(two_img).float().to(self.device)
            self.model(representation)
        print("Ready to predict the 6D force!")

        # wrench range setup for un-normalization
        self.wrench_range = np.array([[-3, 3], [-3, 3], [-12, 0], [-0.2, 0.2], [-0.2, 0.2], [-0.05, 0.05]])
        self.min_wrench = self.wrench_range[:, 0]
        self.max_wrench = self.wrench_range[:, 1]
        self.wrench_span = self.max_wrench - self.min_wrench

    def predict_force(self, representation):
        representation = representation.transpose([2, 0, 1])
        input_batch = 2
        if input_batch == 2:
            two_img = np.array([representation, representation])
            representation = torch.from_numpy(two_img).float().to(self.device)
            with torch.no_grad():
                force = self.model(representation)
                force[force < 0] = 0
                force[force > 1] = 1
                force = force.cpu().numpy()[0]
        else:
            representation = torch.from_numpy(np.array(representation)).float().to(self.device)
            with torch.no_grad():
                force = self.model(representation)
                force[force < 0] = 0
                force[force > 1] = 1
                force = force.cpu().numpy()

        # Unnormalize to get the real-world force values
        real_force = force * self.wrench_span + self.min_wrench

        return force, real_force

if __name__ == '__main__':
    f = open("force_config.yaml", 'r+', encoding='utf-8')
    config = yaml.load(f, Loader=yaml.FullLoader)
    estimator = Estimator(config)
    image = cv2.imread('test.png')
    input_image = np.zeros_like(image)
    input_image[::, ::, 0] = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    normalized_force, real_force = estimator.predict_force(input_image)
    print("Normalized 6D force output:", normalized_force)
    print("Real 6D force output (unnormalized):", real_force)

```


Also on a test image with no force applied, shouldn't the 6D vector of Fx,Fy,Fz,Tx,Ty,Tz all be zero?

This is what I got on a test image with no force applied after camera calibration, sensor calibration and sensor calibration. I also made sure we are using the DenseNet-169 pre-trained values you mentioned in my previous inquiry:

Force estimate on test image with no force applied for sure:

```
use Densenet-169
Ready to predict the 6D force!
Normalized 6D force output: [0.5182026  0.47572857 0.65638834 0.4747542  0.544809   0.46920186]
Real 6D force output (unnormalized): [ 0.10921562 -0.14562857 -4.12333989 -0.01009831  0.01792359 -0.00307981]

```

Why Fz=-4.12333989? Why any of these values nonzero? Shouldn't they all be zero?

This was my test image which no force is applied:

![raw-test](https://github.com/user-attachments/assets/119d5668-d40c-47df-b688-99ac05cce62b)


We are so close to getting this working, thank you for your continued help and support @linchangyi1 ! We appreciate you very much :)

**Comment by linchangyi1** [comment]
Hi @f5ces27er5c767s, the pre-trained model will output non-zero values because it was trained on the dataset collected from my sensor. That's why you need to collect data from your sensor for fine-tuning the model.

**Comment by f5ces27er5c767s** [comment]
@linchangyi1 Ok thanks, so I should expect though those values to be zero though if I collect data from my sensor?

**Comment by linchangyi1** [comment]
If you retrain the model, the output is expected to be zeros, but actually it will not be exactly zero because of the small estimation error.
